lib/uplink/main.go

```diff
@@ -0,0 +1,238 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"io"
+	"time"
+
+	"storj.io/storj/pkg/miniogw"
+	"storj.io/storj/pkg/ranger"
+	"storj.io/storj/pkg/storj"
+)
+
+// An Identity is a parsed leaf cert keypair with a certificate signed chain
+// up to the self-signed CA with the node ID.
+type Identity interface {
+	NodeID() storj.NodeID
+	Key() crypto.PrivateKey
+	Certs() []x509.Certificate
+}
+
+// Caveats could be things like a read-only restriction, a time-bound
+// restriction, a bucket-specific restriction, a path-prefix restriction, a
+// full path restriction, etc.
+type Caveat interface {
+}
+
+// A Macaroon represents an access credential to certain resources
+type Macaroon interface {
+	Serialize() ([]byte, error)
+	Restrict(caveats ...Caveat) Macaroon
+}
+
+type Config struct {
+	// MaxBufferMem controls upload performance and is system-specific
+	MaxBufferMem int
+
+	// These should only be relevant for new files; these values for existing
+	// files should come from the metainfo index. It's unlikely these will ever
+	// change much.
+	EncBlockSize  int
+	MaxInlineSize int
+	SegmentSize   int64
+}
+
+// Uplink represents the main entrypoint to Storj V3. An Uplink connects to
+// a specific Satellite and caches connections and resources, allowing one to
+// create sessions delineated by specific access controls.
+type Uplink struct {
+}
+
+// NewUplink creates a new Uplink
+func NewUplink(identity Identity, satelliteAddr string, cfg Config) *Uplink {
+	panic("TODO")
+}
+
+// A Share is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+type Share struct {
+	Access Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
```

Comment:
maybe the solution to this problem is to have two levels of macaroon - one for project management, and then one for within a specific bucket. then sessions are scoped to a specific bucket maybe?